pkg/resource/deploy/source_eval.go

```diff
@@ -1765,7 +1776,40 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 	}
 	customTimeouts := opts.CustomTimeouts
 
-	additionalSecretOutputs := opts.GetAdditionalSecretOutputs()
+	// Schema-aware changes
+	//
+	// At this point we will consult the resource's schema to determine if any of the properties we'll send onward should
+	// be modified or extended. Presently this includes:
+	//
+	// * Including `Secret` properties in additional secret outputs.
+
+	additionalSecretOutputsSet := mapset.NewSet[string]()
+	additionalSecretOutputsSet.Append(opts.GetAdditionalSecretOutputs()...)
+
+	//var version *semver.Version
+	//if parsedVersion, err := semver.Parse(req.GetVersion()); err == nil {
+	//  version = &parsedVersion
+	//}
+
+	//if typeToken, err := tokens.ParseTypeToken(req.Type); err == nil {
+	//  packageName := typeToken.Package().String()
+	//  if pkgReference, err := rm.schemaLoader.LoadPackageReference(packageName, version); err == nil {
+	//    if resourceSchema, found, err := pkgReference.Resources().Get(req.Type); err == nil && found {
+	//      for _, outputProperty := range resourceSchema.Properties {
+	//        if outputProperty.Secret {
+	//          additionalSecretOutputsSet.Add(outputProperty.Name)
+	//        }
+	//      }
+	//    }
+	//  }
+	//}
+
+	var additionalSecretOutputs []string
+	if additionalSecretOutputsSet.Cardinality() > 0 {
+		additionalSecretOutputs = additionalSecretOutputsSet.ToSlice()
+	} else {
+		additionalSecretOutputs = nil
+	}
```

Comment:
Does this not work?
```suggestion
	additionalSecretOutputs = additionalSecretOutputsSet.ToSlice()
```

Reply:
Alas no, because that will return an empty slice for `nil`, which is not the current behaviour (`nil` stays as `nil`). I'd like to believe that semantically it will make no difference, but for "perfect" backwards compatibility I've opted for this. Happy to go either way though.